src/main/java/org/codeforamerica/messaging/models/Template.java

```diff
@@ -27,19 +32,66 @@ public class Template {
+    @JsonIgnore
     private OffsetDateTime creationTimestamp;
     @UpdateTimestamp
+    @JsonIgnore
     private OffsetDateTime updateTimestamp;
 
-    public void addTemplateVariant(TemplateVariant templateVariant) {
-        this.templateVariants.add(templateVariant);
-        templateVariant.setTemplate(this);
+    public void addTemplateVariant(TemplateVariant templateVariant) throws Exception {
+        if (this.templateVariants.contains(templateVariant)) {
+            throw new Exception("Template variant already exists");
+        } else {
+            this.templateVariants.add(templateVariant);
+            templateVariant.setTemplate(this);
+        }
     }
 
-    public void removeTemplateVariant(TemplateVariant templateVariant) {
-        this.templateVariants.remove(templateVariant);
+    public void updateTemplateVariant(TemplateVariant templateVariant, String body, String subject)
+            throws Exception {
+        if (isTemplateVariantInUse(templateVariant)) {
+            throw new Exception("Cannot update a template variant that is already in use");
+        }
+        templateVariant.setBody(body);
+        templateVariant.setSubject(subject);
+    }
+
+    public void mergeTemplateVariant(TemplateVariant templateVariant) throws Exception {
+        Optional<TemplateVariant> existingTemplateVariant =
+                this.getTemplateVariant(templateVariant.getLanguage(), templateVariant.getTreatment());
+        if (existingTemplateVariant.isPresent()) {
+            this.updateTemplateVariant(
+                    existingTemplateVariant.get(),
+                    templateVariant.getBody(),
+                    templateVariant.getSubject()
+            );
+        } else {
+            this.addTemplateVariant(templateVariant);
+        }
+    }
+
+    public void removeTemplateVariant(String language, String treatment) throws Exception {
+        TemplateVariant templateVariant = this.getTemplateVariant(language, treatment).orElseThrow(NoSuchElementException::new);
+        if (this.getTemplateVariants().size() == 1) {
+            throw new Exception("Cannot delete last variant on template - delete parent template instead");
+        }
+        if (isTemplateVariantInUse(templateVariant)) {
+            throw new Exception("Template variant is currently in use and cannot be deleted");
+        }
         templateVariant.setTemplate(null);
+        this.getTemplateVariants().remove(templateVariant);
+    }
+
+    public Optional<TemplateVariant> getTemplateVariant(String language, String treatment) {
+        return this.getTemplateVariants().stream()
+                .filter(templateVariant -> templateVariant.getLanguage().equals(language))
+                .filter(templateVariant -> templateVariant.getTreatment().equals(treatment))
+                .findAny();
+    }
+
+    public static boolean isTemplateVariantInUse(TemplateVariant templateVariant) {
+        return templateVariant.getMessages().size() > 0;
```

Comment:
As this would reify all the messages it may be better to use query derivation on the message repository using 
`count…By`.


<br class="Apple-interchange-newline">